pkg/suggestion/v1beta1/optuna/service.py

```diff
@@ -0,0 +1,178 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from collections import defaultdict
+import threading
+
+import optuna
+
+from pkg.apis.manager.v1beta1.python import api_pb2
+from pkg.apis.manager.v1beta1.python import api_pb2_grpc
+from pkg.suggestion.v1beta1.internal.constant import INTEGER, DOUBLE, CATEGORICAL, DISCRETE, MAX_GOAL
+from pkg.suggestion.v1beta1.internal.search_space import HyperParameterSearchSpace
+from pkg.suggestion.v1beta1.internal.trial import Trial, Assignment
+from pkg.suggestion.v1beta1.internal.base_health_service import HealthServicer
+
+
+class OptunaService(api_pb2_grpc.SuggestionServicer, HealthServicer):
+
+    def __init__(self):
+        super(OptunaService, self).__init__()
+        self.study = None
+        self.search_space = None
+        self.recorded_trial_names = set()
+        self.assignments_to_optuna_number = defaultdict(list)
+        self.lock = threading.Lock()
+
+    def GetSuggestions(self, request, context):
+        """
+        Main function to provide suggestion.
+        """
+        with self.lock:
+            if self.study is None:
+                self.search_space = HyperParameterSearchSpace.convert(request.experiment)
+                self.study = self._create_study(request.experiment.spec.algorithm, self.search_space)
+
+            trials = Trial.convert(request.trials)
+
+            if len(trials) != 0:
+                self._tell(trials)
+            list_of_assignments = self._ask(request.request_number)
+
+            return api_pb2.GetSuggestionsReply(
+                parameter_assignments=Assignment.generate(list_of_assignments)
+            )
+
+    def _create_study(self, algorithm_spec, search_space):
+        sampler = self._create_sampler(algorithm_spec)
+        direction = "maximize" if search_space.goal == MAX_GOAL else "minimize"
+
+        study = optuna.create_study(sampler=sampler, direction=direction)
+
+        return study
+
+    def _create_sampler(self, algorithm_spec):
+        name = algorithm_spec.algorithm_name
+        settings = {s.name:s.value for s in algorithm_spec.algorithm_settings}
+
+        if name == "tpe" or name == "multivariate-tpe":
+            kwargs = {}
+            for k, v in settings.items():
+                if k == "startup_trials":
+                    kwargs["n_startup_trials"] = int(v)
+                elif k == "ei_candidates":
+                    kwargs["n_ei_candidates"] = int(v)
+                elif k == "random_state":
+                    kwargs["seed"] = int(v)
+                else:
+                    raise ValueError("Unknown name for {}: {}".format(name, k))
+
+            kwargs["multivariate"] = name == "multivariate-tpe"
+
+            sampler = optuna.samplers.TPESampler(**kwargs)
```

Comment:
@g-votte I think it's reasonable to set `kwargs["constant_liar"] = True` by default. Because you know, Katib allows us to run distributed optimization easily. What do you think?

Reply:
Or we can provide an algorithm setting to set `constant_liar=True`.

Reply:
Good catch. There is less reason to disable constant liar especially with parallel optimization.
Added a line to turn on the option by default.
https://github.com/kubeflow/katib/pull/1613/commits/1d8188534de9e10a028fa7253f33cbf5de7eea87